### PR TITLE
MHub / GC - Add the Node21 baseline model

### PR DESCRIPTION
**Pull request by silvandeleemput**
This PR adds the GC STOIC baseline model (from the [STOIC2021 challenge](https://stoic2021.grand-challenge.org/)) to MHub.
GitHub Repo: https://github.com/node21challenge/node21_detection_baseline
GC page: https://node21.grand-challenge.org/

### Algorithm I/O
  * Input should be a Chest X-Ray image (MHA) (This MHub implementation expects a Dicom which is converted to MHA using panimg MhaConverter)
  * Outputs are:
    * A JSON file with the bounding boxes and probabilities for the nodules
  * Test DICOM images for this algorithm can be downloaded from IDC:
 ```
# To download the files in this manifest, first install s5cmd (https://github.com/peak/s5cmd),
# then run the following command:
# s5cmd --no-sign-request --endpoint-url https://storage.googleapis.com run file_manifest_gcs.s5cmd
cp s3://public-datasets-idc/8fb5dfd4-0bd1-4dc2-a8a7-7a867c40450e/* .
cp s3://public-datasets-idc/3c0bd6df-1772-4bb2-abe5-570665311253/* .
```

### Caveats
* I added the `CR = Meta(mode="CR")       # CR	Computed Radiography` label, which should move to `mhubio/core/templates.py`, not really sure if this is the best label for Chest X-Rays (CXR)
* ~The PR target is `main`, but should be something like `m-gc-node21-baseline`~
* ~In the Dockerfile the MHub model repo integration is currently marked `TODO` and should be uncommented/edited~
* ~Added a TODO to the `MHAConverter` since it doesn't have the `panimg` backend yet (without this backend it fails for my test image), and it uses a `run.py` script to use the panimg mhaconverter.~

### CODE REVIEW

**Comment by silvandeleemput** [comment]
Updated this PR:
* Dockerfile
  * Use mhubio.run as entrypoint
  * Fixed commit for node21 repo
  * Updated TODO on MHub model code integration
* default.yml
  * Changed mhaconverter backend to panimg
* Node21BaselineRunner.py
  * Added comments
* Removed
  * run.py
  * PanImgConverters.py

**Comment by LennyN95** [comment]
_**NOTE**: This PR proposes a model currently using a dynamic length output. We are working on a method to export dynamic length outputs and will likely introduce a `@IO.Out.Data.Many` decorator (similar to our `@IO.Output` -> `@IO.Outputs` decorators)._

File: models/gc_node21_baseline/utils/Node21BaselineRunner.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+# TODO should move to mhubio/core/templates.py
+CR = Meta(mode="CR")  # CR	Computed Radiography
+
+
+class Node21BaselineRunner(Module):
+
+    @IO.Instance()
+    @IO.Input('in_data', 'mha|nrrd|nifti:mod=cr', the='input chest X-Ray')
+    @IO.Output('out_data', 'nodules.json', 'json:model=Node21Baseline', 'in_data', the='Node21 baseline nodule prediction')
```
Comment (LennyN95) [inline models/gc_node21_baseline/utils/Node21BaselineRunner.py]: In addition to the original file output (which should always be accessible to reproduce the original pipeline as-is) we need to include the value-based findings via the `@IO.OutputData(s)` decorator to make these values accessible to our ReportExporter (and thereby allowing a standardized output structure).

Comment (silvandeleemput) [reply models/gc_node21_baseline/utils/Node21BaselineRunner.py]: That seems fine, does it have to be implemented using the `@IO.Out.Data.Many` decorator or is `@IO.OutputData(s)` sufficient? Do you maybe have an example of how it works?

Comment (silvandeleemput) [reply models/gc_node21_baseline/utils/Node21BaselineRunner.py]: A small example of the JSON output by the model:

```json
{
    "type": "Multiple 2D bounding boxes",
    "boxes": [
        {
            "corners": [
                [
                    274.9,
                    224.49,
                    0
                ],
                [
                    258.17,
                    224.49,
                    0
                ],
                [
                    258.17,
                    206.72,
                    0
                ],
                [
                    274.9,
                    206.72,
                    0
                ]
            ],
            "probability": 0.44
        },

        {
            "corners": [
                [
                    315.46,
                    266.54,
                    0
                ],
                [
                    292.71,
                    266.54,
                    0
                ],
                [
                    292.71,
                    239.46,
                    0
                ],
                [
                    315.46,
                    239.46,
                    0
                ]
            ],
            "probability": 0.08
        },
    ],
    "version": {
        "major": 1,
        "minor": 0
    }
}

```

Comment (LennyN95) [reply models/gc_node21_baseline/utils/Node21BaselineRunner.py]: The second example here will be the most relevant to your use-case: https://github.com/MHubAI/documentation/blob/main/documentation/mhubio/how_to_write_an_mhubio_module.md#dynamic-output-1n

If you want to export the bounding box plus probability, you must define 4*3+1 value outputs. 
This is one of the edge cases, so instead suggest we discuss the format. n MHub-specific `Coordinate` datatype with additional parameters could be a solution. The problem I see with plain floating point coordinates is, that when comparing multiple algorithms the individually used reference frames might be different and the coordinates thus are not directly comparable.

So for now, a single ValueOutput for the probability risk can be used.

In the near future, we'll rename all decorators to a more straightforward naming schema. You can stick to the old syntax for now, we'll patch all implemented models once this becomes effective.
```
@IO.Input       -> @IO.In.File.One
@IO.Inputs      -> @IO.In.File.Many
@IO.Output      -> @IO.Out.File.One
@IO.Outputs     -> @IO.Out.File.Many
@IO.OutputData  -> @IO.Out.Data.One
@IO.OutputDatas -> @IO.Out.Data.Many
```

Comment (silvandeleemput) [reply models/gc_node21_baseline/utils/Node21BaselineRunner.py]: Thanks for the explanation! I stick to your suggestion for a single ValueOutput for now.

Comment (LennyN95) [reply models/gc_node21_baseline/utils/Node21BaselineRunner.py]: I was thinking about it. Let's do two. One for the probability, type float. And one for the bounding box type str. 
Then add `Meta(format='json')` besides any other metadata to the bounding box type. 
Encoding the bounding boxes as JSON strings is yet flexible and can be fully explained in the description. 

Some questions:
In what reference frame are the coordinates? 
Does any re-sampling happen of the image?

Comment (silvandeleemput) [reply models/gc_node21_baseline/utils/Node21BaselineRunner.py]: I implemented your suggestion. 

Regarding the questions:
Looking at the source code there appears to be no re-sampling of the input image. 
The bounding box coordinates are apparently in mm counting from the top-left corner of the image.

**Comment by silvandeleemput** [comment]
[gc_node21_baseline_output.zip](https://github.com/MHubAI/models/files/15024198/gc_node21_baseline_output.zip)

**Comment by silvandeleemput** [comment]
/test

```yaml
sample:
  idc_version: 17.0
  data:
    - SeriesInstanceUID: 1.3.6.1.4.1.14519.5.2.1.6279.6001.210102868760281756294235082201
      aws_url: s3://idc-open-data/b1c859c3-7471-43e7-9882-0bb5420ec18a/*
      path: dicom

reference:
  url: https://github.com/MHubAI/models/files/15024198/gc_node21_baseline_output.zip
```

---

<details><summary>Test Results (24.04.18_22.31.12_E3krCTzQUg)</summary>

```yaml
id: e8377a1b-338f-4fab-8647-fe4b72a250be
date: '2024-04-18 22:36:27'
checked_files:
- file: nodules.json
  path: /app/test/src/1.3.6.1.4.1.14519.5.2.1.6279.6001.210102868760281756294235082201/nodules.json
  checks:
  - checker: DataFileCheck
    notes:
    - label: Value Match
      description: These keys have identical values
      info: type,boxes.[0].corners.[0].[0],boxes.[0].corners.[0].[1],boxes.[0].corners.[0].[2],boxes.[0].corners.[1].[0],boxes.[0].corners.[1].[1],boxes.[0].corners.[1].[2],boxes.[0].corners.[2].[0],boxes.[0].corners.[2].[1],boxes.[0].corners.[2].[2],boxes.[0].corners.[3].[0],boxes.[0].corners.[3].[1],boxes.[0].corners.[3].[2],boxes.[0].probability,boxes.[1].corners.[0].[0],boxes.[1].corners.[0].[1],boxes.[1].corners.[0].[2],boxes.[1].corners.[1].[0],boxes.[1].corners.[1].[1],boxes.[1].corners.[1].[2]
        (+ 31 more)
    findings:
    - label: Value similar
      description: Value of key 'boxes.[2].corners.[0].[0]' is similar in source file
      subpath: boxes.[2].corners.[0].[0]
      info:
        scale: 10
        precision: -1
        src: 56.17
        ref: 52.17
    - label: Value similar
      description: Value of key 'boxes.[2].corners.[0].[1]' is similar in source file
      subpath: boxes.[2].corners.[0].[1]
      info:
        scale: 100
        precision: -1
        src: 210.43
        ref: 259.47
    - label: Value similar
      description: Value of key 'boxes.[2].corners.[1].[0]' is similar in source file
      subpath: boxes.[2].corners.[1].[0]
      info:
        scale: 1
        precision: -1
        src: 48.52
        ref: 31.0
    - label: Value similar
      description: Value of key 'boxes.[2].corners.[1].[1]' is similar in source file
      subpath: boxes.[2].corners.[1].[1]
      info:
        scale: 100
        precision: -1
        src: 210.43
        ref: 259.47
    - label: Value similar
      description: Value of key 'boxes.[2].corners.[2].[0]' is similar in source file
      subpath: boxes.[2].corners.[2].[0]
      info:
        scale: 1
        precision: -1
        src: 48.52
        ref: 31.0
    - label: Value similar
      description: Value of key 'boxes.[2].corners.[2].[1]' is similar in source file
      subpath: boxes.[2].corners.[2].[1]
      info:
        scale: 100
        precision: -1
        src: 201.47
        ref: 237.66
    - label: Value similar
      description: Value of key 'boxes.[2].corners.[3].[0]' is similar in source file
      subpath: boxes.[2].corners.[3].[0]
      info:
        scale: 10
        precision: -1
        src: 56.17
        ref: 52.17
    - label: Value similar
      description: Value of key 'boxes.[2].corners.[3].[1]' is similar in source file
      subpath: boxes.[2].corners.[3].[1]
      info:
        scale: 100
        precision: -1
        src: 201.47
        ref: 237.66
    - label: Value similar
      description: Value of key 'boxes.[3].corners.[0].[0]' is similar in source file
      subpath: boxes.[3].corners.[0].[0]
      info:
        scale: 10
        precision: -1
        src: 52.16
        ref: 56.17
    - label: Value similar
      description: Value of key 'boxes.[3].corners.[0].[1]' is similar in source file
      subpath: boxes.[3].corners.[0].[1]
      info:
        scale: 100
        precision: -1
        src: 259.48
        ref: 210.43
    - label: Value similar
      description: Value of key 'boxes.[3].corners.[1].[0]' is similar in source file
      subpath: boxes.[3].corners.[1].[0]
      info:
        scale: 1
        precision: -1
        src: 31.0
        ref: 48.52
    - label: Value similar
      description: Value of key 'boxes.[3].corners.[1].[1]' is similar in source file
      subpath: boxes.[3].corners.[1].[1]
      info:
        scale: 100
        precision: -1
        src: 259.48
        ref: 210.43
    - label: Value similar
      description: Value of key 'boxes.[3].corners.[2].[0]' is similar in source file
      subpath: boxes.[3].corners.[2].[0]
      info:
        scale: 1
        precision: -1
        src: 31.0
        ref: 48.52
    - label: Value similar
      description: Value of key 'boxes.[3].corners.[2].[1]' is similar in source file
      subpath: boxes.[3].corners.[2].[1]
      info:
        scale: 100
        precision: -1
        src: 237.66
        ref: 201.47
    - label: Value similar
      description: Value of key 'boxes.[3].corners.[3].[0]' is similar in source file
      subpath: boxes.[3].corners.[3].[0]
      info:
        scale: 10
        precision: -1
        src: 52.16
        ref: 56.17
    - label: Value similar
      description: Value of key 'boxes.[3].corners.[3].[1]' is similar in source file
      subpath: boxes.[3].corners.[3].[1]
      info:
        scale: 100
        precision: -1
        src: 237.66
        ref: 201.47
    - label: Value similar
      description: Value of key 'boxes.[4].probability' is similar in source file
      subpath: boxes.[4].probability
      info:
        scale: 1
        precision: 1
        src: 0.06
        ref: 0.05
  - checker: SizeCheck
summary:
  files_missing: 0
  files_extra: 0
  checks:
    DataFileCheck:
      files: 1
      findings:
        Value similar: 17
    SizeCheck:
      files: 1
conclusion: false
```

</details>

---

<details><summary>Test Results (24.04.25_08.24.29_0ZZ5RYzDlk)</summary>

```yaml
id: 12aafd5a-a277-4f27-a5c7-708d5670cf91
date: '2024-04-25 08:30:39'
checked_files:
- file: nodules.json
  path: /app/test/src/1.3.6.1.4.1.14519.5.2.1.6279.6001.210102868760281756294235082201/nodules.json
  checks:
  - checker: DataFileCheck
    notes:
    - label: Value Match
      description: These keys have identical values
      info: type,boxes.[0].corners.[0].[0],boxes.[0].corners.[0].[1],boxes.[0].corners.[0].[2],boxes.[0].corners.[1].[0],boxes.[0].corners.[1].[1],boxes.[0].corners.[1].[2],boxes.[0].corners.[2].[0],boxes.[0].corners.[2].[1],boxes.[0].corners.[2].[2],boxes.[0].corners.[3].[0],boxes.[0].corners.[3].[1],boxes.[0].corners.[3].[2],boxes.[0].probability,boxes.[1].corners.[0].[0],boxes.[1].corners.[0].[1],boxes.[1].corners.[0].[2],boxes.[1].corners.[1].[0],boxes.[1].corners.[1].[1],boxes.[1].corners.[1].[2]
        (+ 43 more)
    findings:
    - label: Value similar
      description: Value of key 'boxes.[2].corners.[0].[0]' is similar in source file
      subpath: boxes.[2].corners.[0].[0]
      info:
        scale: 1
        precision: 1
        src: 52.16
        ref: 52.17
    - label: Value similar
      description: Value of key 'boxes.[2].corners.[0].[1]' is similar in source file
      subpath: boxes.[2].corners.[0].[1]
      info:
        scale: 1
        precision: 1
        src: 259.48
        ref: 259.47
    - label: Value similar
      description: Value of key 'boxes.[2].corners.[1].[1]' is similar in source file
      subpath: boxes.[2].corners.[1].[1]
      info:
        scale: 1
        precision: 1
        src: 259.48
        ref: 259.47
    - label: Value similar
      description: Value of key 'boxes.[2].corners.[3].[0]' is similar in source file
      subpath: boxes.[2].corners.[3].[0]
      info:
        scale: 1
        precision: 1
        src: 52.16
        ref: 52.17
    - label: Value similar
      description: Value of key 'boxes.[4].probability' is similar in source file
      subpath: boxes.[4].probability
      info:
        scale: 1
        precision: 1
        src: 0.06
        ref: 0.05
  - checker: SizeCheck
summary:
  files_missing: 0
  files_extra: 0
  checks:
    DataFileCheck:
      files: 1
      findings:
        Value similar: 5
    SizeCheck:
      files: 1
conclusion: false
```

</details>

**Comment by silvandeleemput** [comment]
/review

We updated the meta.json file, we implemented the suggestion for the OutputDatas (probabilities, and bounding boxes as JSON), and we have wrapped the algorithm/model in a CLI to avoid stdout capture issues. It should be as good as done, please have a look.

**Review by LennyN95 — APPROVED** [review]

**Review by LennyN95 — COMMENTED** [review]
To keep the structure clean, the `utils` folder ideally only contains any MHub-IO related utilities (e.g., custom Modules such as the Node21BaselineRunner) but no model implementation code. All model implementation should go into an `src` folder, all additional resources into an `res` folder. 

Making such a change is purely organizational (although for the utils folder, it is somewhat meaningful as we parse this folder to detect available Modules).

However, in case you have to make any code adjustments and we have to redo the test, I'd ask for this change.

**Review by LennyN95 — CHANGES_REQUESTED** [review]
The test failed and showed (high) variation in the bounding boxes. This needs to be investigated before we can proceed.

**Comment by silvandeleemput** [comment]
I investigated this issue and found the root cause for the large bounding box differences.

Rerunning on my machine consistently gave me the same output as I have uploaded before.
I have 5 detected nodules with each 4 corners (3 values x,y,z=0) and a probability.

Comparing these with your test results I found:
- The number of detected bounding boxes matches the number I found on my side (5). This seems ok.
- All found probabilities match pretty closely. (some small deviations are expected because of GPU hardware/driver differences and rounding errors). These seem fine.
- Bounding boxes 0, 1, and 4 seem fine and the same as what I found.
- Bounding boxes 2 & 3 are swapped, having the same probability! Giving rise to the big found differences. It seems to be a sorting issue.

I can fix this by applying a better sorting algorithm after the output is generated to produce a more consistent output JSON file. This can be added to the cli.py and go together with a small refactor of the cli.py file to a /src directory. This should be under `/models/gc_node21_baseline/src/` then ?

**Comment by silvandeleemput** [comment]
@LennyN95 alright, I have added the improved sorting, moved the cli.py to a src directory, and updated the model version to 1.1, both in the meta.json and the Dockerfile. The previous test can be redone and should produce better results because the sorting issue is fixed. This model should be as good as done.

**Comment by silvandeleemput** [comment]
I looked at the new result generated after adding the sorting fix, and the sorting works now: the same bounding boxes are found and sorted in the correct order. The found differences this time around are only small rounding errors for both the coordinates of the bounding boxes and for the probabilities, which are expected. This result is acceptable.

**Review by LennyN95 — APPROVED** [review]
Test passed (rounding errors expected, see comment from the model contributor).

**Pull request merged**